File: game/event.go
package game

import (
	"github.com/BigJk/end_of_eden/internal/lua/luhelp"
)

// EventChoice represents a possible choice in the Event.
type EventChoice struct {
	Description   string
	DescriptionFn luhelp.OwnedCallback
	Callback      luhelp.OwnedCallback
}

// Event represents a encounter-able event.
type Event struct {
	ID          string
	Name        string
	Description string
	Tags        []string
	Choices     []EventChoice
	OnEnter     luhelp.OwnedCallback
	OnEnd       luhelp.OwnedCallback
	Test        luhelp.OwnedCallback
	BaseGame    bool
}

func (e Event) IsNone() bool {
	return len(e.ID) == 0
}

File: assets/scripts/definitions/event.lua
---@meta

---@class event_on_enter_ctx
---@field type_id type_id

---@class event_choice_ctx
---@field type_id type_id
---@field choice number

---EventChoice represents a possible choice in the Event.
---@class event_choice
---@field description? string
---@field description_fn? fun():nil|string
---@field callback fun(ctx:event_choice_ctx):next_game_state|nil

---Event represents a encounter-able event.
---@class event
---@field id? string
---@field name string
---@field description string
---@field choices event_choice[]
---@field on_enter? fun(ctx:event_on_enter_ctx):nil
---@field on_end fun(ctx:event_choice_ctx):next_game_state|nil
---@field test? fun():nil|string
---@field base_game? boolean